Libraries/curl/8.8.0/share/man/man1/curl.1
.\" **************************************************************************
.\" *                                  _   _ ____  _
.\" *  Project                     ___| | | |  _ \| |
.\" *                             / __| | | | |_) | |
.\" *                            | (__| |_| |  _ <| |___
.\" *                             \___|\___/|_| \_\_____|
.\" *
.\" * Copyright (C) Daniel Stenberg, <daniel@haxx.se>, et al.
.\" *
.\" * This software is licensed as described in the file COPYING, which
.\" * you should have received as part of this distribution. The terms
.\" * are also available at https://curl.se/docs/copyright.html.
.\" *
.\" * You may opt to use, copy, modify, merge, publish, distribute and/or sell
.\" * copies of the Software, and permit persons to whom the Software is
.\" * furnished to do so, under the terms of the COPYING file.
.\" *
.\" * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
.\" * KIND, either express or implied.
.\" *
.\" * SPDX-License-Identifier: curl
.\" *
.\" **************************************************************************
.\"
.\" DO NOT EDIT. Generated by the curl project managen man page generator.
.\"
.TH curl 1 "2024-05-22" "curl 8.8.0" "curl Manual"
.SH NAME
curl \- transfer a URL
.SH SYNOPSIS
\fBcurl [options / URLs]\fP
.SH DESCRIPTION
\fBcurl\fP is a tool for transferring data from or to a server using URLs. It
supports these protocols: DICT, FILE, FTP, FTPS, GOPHER, GOPHERS, HTTP, HTTPS,
IMAP, IMAPS, LDAP, LDAPS, MQTT, POP3, POP3S, RTMP, RTMPS, RTSP, SCP, SFTP,
SMB, SMBS, SMTP, SMTPS, TELNET, TFTP, WS and WSS.

curl is powered by libcurl for all transfer\-related features. See
\fIlibcurl(3)\fP for details.
.SH URL
The URL syntax is protocol\-dependent. You find a detailed description in
RFC 3986.

If you provide a URL without a leading \fBprotocol://\fP scheme, curl guesses
what protocol you want. It then defaults to HTTP but assumes others based on
often\-used hostname prefixes. For example, for hostnames starting with "ftp."
curl assumes you want FTP.

You can specify any amount of URLs on the command line. They are fetched in a
sequential manner in the specified order unless you use \fI\-Z, \-\-parallel\fP. You can
specify command line options and URLs mixed and in any order on the command
line.

curl attempts to reuse connections when doing multiple transfers, so that
getting many files from the same server do not use multiple connects and setup
handshakes. This improves speed. Connection reuse can only be done for URLs
specified for a single command line invocation and cannot be performed between
separate curl runs.

Provide an IPv6 zone id in the URL with an escaped percentage sign. Like in
.nf

\&"http://[fe80::3%25eth0]/"
.fi

Everything provided on the command line that is not a command line option or
its argument, curl assumes is a URL and treats it as such.
.SH GLOBBING
You can specify multiple URLs or parts of URLs by writing lists within braces
or ranges within brackets. We call this "globbing".

Provide a list with three different names like this:
.nf

\&"http://site.{one,two,three}.com"
.fi

Do sequences of alphanumeric series by using [] as in:
.nf

\&"ftp://ftp.example.com/file[1\-100].txt"
.fi

With leading zeroes:
.nf

\&"ftp://ftp.example.com/file[001\-100].txt"
.fi

With letters through the alphabet:
.nf

\&"ftp://ftp.example.com/file[a\-z].txt"
.fi

Nested sequences are not supported, but you can use several ones next to each
other:
.nf

\&"http://example.com/archive[1996\-1999]/vol[1\-4]/part{a,b,c}.html"
.fi

You can specify a step counter for the ranges to get every Nth number or
letter:
.nf

\&"http://example.com/file[1\-100:10].txt"

\&"http://example.com/file[a\-z:2].txt"
.fi

When using [] or {} sequences when invoked from a command line prompt, you
probably have to put the full URL within double quotes to avoid the shell from
interfering with it. This also goes for other characters treated special, like
for example \(aq&\(aq, \(aq?\(aq and \(aq*\(aq.

Switch off globbing with \fI\-g, \-\-globoff\fP.
.SH VARIABLES
curl supports command line variables (added in 8.3.0). Set variables with
\fI\-\-variable\fP name=content or \fI\-\-variable\fP name@file (where "file" can be stdin if
set to a single dash (\-)).

Variable contents can be expanded in option parameters using "{{name}}" if the
option name is prefixed with "\fI\-\-expand\-\fP". This gets the contents of the
variable "name" inserted, or a blank if the name does not exist as a
variable. Insert "{{" verbatim in the string by prefixing it with a backslash,
like "\\{{".

You an access and expand environment variables by first importing them. You
can select to either require the environment variable to be set or you can
provide a default value in case it is not already set. Plain \fI\-\-variable\fP %name
imports the variable called \(aqname\(aq but exits with an error if that environment
variable is not already set. To provide a default value if it is not set, use
\fI\-\-variable\fP %name=content or \fI\-\-variable\fP %name@content.

Example. Get the USER environment variable into the URL, fail if USER is not
set:
.nf

-\-variable \(aq%USER\(aq
-\-expand\-url = "https://example.com/api/{{USER}}/method"
.fi

When expanding variables, curl supports a set of functions that can make the
variable contents more convenient to use. It can trim leading and trailing
white space with "trim", it can output the contents as a JSON quoted string
with "json", URL encode the string with "url" or base64 encode it with "b64".
To apply functions to a variable expansion, add them colon separated to the
right side of the variable. Variable content holding null bytes that are not
encoded when expanded cause error.

Example: get the contents of a file called $HOME/.secret into a variable
called "fix". Make sure that the content is trimmed and percent\-encoded when
sent as POST data:
.nf

-\-variable %HOME
-\-expand\-variable fix@{{HOME}}/.secret
-\-expand\-data "{{fix:trim:url}}"
https://example.com/
.fi

Command line variables and expansions were added in 8.3.0.
.SH OUTPUT
If not told otherwise, curl writes the received data to stdout. It can be
instructed to instead save that data into a local file, using the \fI\-o, \-\-output\fP or
\fI\-O, \-\-remote\-name\fP options. If curl is given multiple URLs to transfer on the
command line, it similarly needs multiple options for where to save them.

curl does not parse or otherwise "understand" the content it gets or writes as
output. It does no encoding or decoding, unless explicitly asked to with
dedicated command line options.
.SH PROTOCOLS
curl supports numerous protocols, or put in URL terms: schemes. Your
particular build may not support them all.
.IP DICT
Lets you lookup words using online dictionaries.
.IP FILE
Read or write local files. curl does not support accessing file:// URL
remotely, but when running on Microsoft Windows using the native UNC approach
works.
.IP FTP(S)
curl supports the File Transfer Protocol with a lot of tweaks and levers. With
or without using TLS.
.IP GOPHER(S)
Retrieve files.
.IP HTTP(S)
curl supports HTTP with numerous options and variations. It can speak HTTP
version 0.9, 1.0, 1.1, 2 and 3 depending on build options and the correct
command line options.
.IP IMAP(S)
Using the mail reading protocol, curl can download emails for you. With or
without using TLS.
.IP LDAP(S)
curl can do directory lookups for you, with or without TLS.
.IP MQTT
curl supports MQTT version 3. Downloading over MQTT equals subscribe to a
topic while uploading/posting equals publish on a topic. MQTT over TLS is not
supported (yet).
.IP POP3(S)
Downloading from a pop3 server means getting a mail. With or without using
TLS.
.IP RTMP(S)
The \fBRealtime Messaging Protocol\fP is primarily used to serve streaming media
and curl can download it.
.IP RTSP
curl supports RTSP 1.0 downloads.
.IP SCP
curl supports SSH version 2 scp transfers.
.IP SFTP
curl supports SFTP (draft 5) done over SSH version 2.
.IP SMB(S)
curl supports SMB version 1 for upload and download.
.IP SMTP(S)
Uploading contents to an SMTP server means sending an email. With or without
TLS.
.IP TELNET
Fetching a telnet URL starts an interactive session where it sends what it
reads on stdin and outputs what the server sends it.
.IP TFTP
curl can do TFTP downloads and uploads.
.SH PROGRESS METER
curl normally displays a progress meter during operations, indicating the
amount of transferred data, transfer speeds and estimated time left, etc. The
progress meter displays the transfer rate in bytes per second. The suffixes
(k, M, G, T, P) are 1024 based. For example 1k is 1024 bytes. 1M is 1048576
bytes.

curl displays this data to the terminal by default, so if you invoke curl to
do an operation and it is about to write data to the terminal, it \fIdisables\fP
the progress meter as otherwise it would mess up the output mixing progress
meter and response data.

If you want a progress meter for HTTP POST or PUT requests, you need to
redirect the response output to a file, using shell redirect (>), \fI\-o, \-\-output\fP
or similar.

This does not apply to FTP upload as that operation does not spit out any
response data to the terminal.

If you prefer a progress bar instead of the regular meter, \fI\-#, \-\-progress\-bar\fP is
your friend. You can also disable the progress meter completely with the
\fI\-s, \-\-silent\fP option.
.SH VERSION
This man page describes curl 8.8.0. If you use a later version, chances
are this man page does not fully document it. If you use an earlier version,
this document tries to include version information about which specific
version that introduced changes.

You can always learn which the latest curl version is by running
.nf

curl https://curl.se/info
.fi

The online version of this man page is always showing the latest incarnation:
https://curl.se/docs/manpage.html
.SH OPTIONS
Options start with one or two dashes. Many of the options require an
additional value next to them. If provided text does not start with a dash, it
is presumed to be and treated as a URL.

The short "single\-dash" form of the options, \-d for example, may be used with
or without a space between it and its value, although a space is a recommended
separator. The long double\-dash form, \fI\-d, \-\-data\fP for example, requires a space
between it and its value.

Short version options that do not need any additional values can be used
immediately next to each other, like for example you can specify all the
options \fI\-O\fP, \fI\-L\fP and \fI\-v\fP at once as \fI\-OLv\fP.

In general, all boolean options are enabled with \--\fBoption\fP and yet again
disabled with \--\fBno\-\fPoption. That is, you use the same option name but
prefix it with "no\-". However, in this list we mostly only list and show the
-\-\fBoption\fP version of them.

When \fI\-:, \-\-next\fP is used, it resets the parser state and you start again with a
clean option state, except for the options that are global. Global options
retain their values and meaning even after \fI\-:, \-\-next\fP.

The following options are global: \fI\-\-fail\-early\fP, \fI\-\-libcurl\fP, \fI\-\-parallel\-immediate\fP, \fI\-Z, \-\-parallel\fP, \fI\-#, \-\-progress\-bar\fP, \fI\-\-rate\fP, \fI\-S, \-\-show\-error\fP, \fI\-\-stderr\fP, \fI\-\-styled\-output\fP, \fI\-\-trace\-ascii\fP, \fI\-\-trace\-config\fP, \fI\-\-trace\-ids\fP, \fI\-\-trace\-time\fP, \fI\-\-trace\fP and \fI\-v, \-\-verbose\fP.
.IP "\-\-abstract\-unix\-socket <path>"
(HTTP) Connect through an abstract Unix domain socket, instead of using the network.
Note: netstat shows the path of an abstract socket prefixed with "@", however
the <path> argument should not have this leading character.

If --abstract-unix-socket is provided several times, the last set value is used.

Example:
.nf
 curl --abstract-unix-socket socketpath https://example.com
.fi

See also \fI\-\-unix\-socket\fP. Added in 7.53.0.
.IP "\-\-alt\-svc <filename>"
(HTTPS) Enable the alt\-svc parser. If the filename points to an existing alt\-svc cache
file, that gets used. After a completed transfer, the cache is saved to the
filename again if it has been modified.

Specify a "" filename (zero length) to avoid loading/saving and make curl just
handle the cache in memory.

If this option is used several times, curl loads contents from all the
files but the last one is used for saving.

--alt-svc can be used several times in a command line

Example:
.nf
 curl --alt-svc svc.txt https://example.com
.fi

See also \fI\-\-resolve\fP and \fI\-\-connect\-to\fP. Added in 7.64.1.
.IP "\-\-anyauth"
(HTTP) Figure out authentication method automatically, and use the most secure one
the remote site claims to support. This is done by first doing a request and
checking the response\-headers, thus possibly inducing an extra network
round\-trip. This option is used instead of setting a specific authentication
method, which you can do with \fI\-\-basic\fP, \fI\-\-digest\fP, \fI\-\-ntlm\fP, and \fI\-\-negotiate\fP.

Using \fI\-\-anyauth\fP is not recommended if you do uploads from stdin, since it may
require data to be sent twice and then the client must be able to rewind. If
the need should arise when uploading from stdin, the upload operation fails.

Used together with \fI\-u, \-\-user\fP.

Providing --anyauth multiple times has no extra effect.

Example:
.nf
 curl --anyauth --user me:pwd https://example.com
.fi

See also \fI\-\-proxy\-anyauth\fP, \fI\-\-basic\fP and \fI\-\-digest\fP.
.IP "\-a, \-\-append"
(FTP SFTP) When used in an upload, this option makes curl append to the target file
instead of overwriting it. If the remote file does not exist, it is
created. Note that this flag is ignored by some SFTP servers (including
OpenSSH).

Providing --append multiple times has no extra effect.
Disable it again with \-\-no-append.

Example:
.nf
 curl --upload-file local --append ftp://example.com/
.fi

See also \fI-r, \-\-range\fP and \fI-C, \-\-continue\-at\fP.
.IP "\-\-aws\-sigv4 <provider1[:prvdr2[:reg[:srv]]]>"
(HTTP) Use AWS V4 signature authentication in the transfer.

The provider argument is a string that is used by the algorithm when creating
outgoing authentication headers.

The region argument is a string that points to a geographic area of
a resources collection (region\-code) when the region name is omitted from
the endpoint.

The service argument is a string that points to a function provided by a cloud
(service\-code) when the service name is omitted from the endpoint.

If --aws-sigv4 is provided several times, the last set value is used.

Example:
.nf
 curl --aws-sigv4 "aws:amz:us-east-2:es" --user "key:secret" https://example.com
.fi

See also \fI\-\-basic\fP and \fI-u, \-\-user\fP. Added in 7.75.0.
.IP "\-\-basic"
(HTTP) Use HTTP Basic authentication with the remote host. This method is the default
and this option is usually pointless, unless you use it to override a
previously set option that sets a different authentication method (such as
\fI\-\-ntlm\fP, \fI\-\-digest\fP, or \fI\-\-negotiate\fP).

Used together with \fI\-u, \-\-user\fP.

Providing --basic multiple times has no extra effect.

Example:
.nf
 curl -u name:password --basic https://example.com
.fi

See also \fI\-\-proxy\-basic\fP.
.IP "\-\-ca\-native"
(TLS) Use the CA store from the native operating system to verify the peer. By
default, curl otherwise uses a CA store provided in a single file or
directory, but when using this option it interfaces the operating system\(aqs own
vault.

This option works for curl on Windows when built to use OpenSSL, wolfSSL
(added in 8.3.0) or GnuTLS (added in 8.5.0). When curl on Windows is built to
use Schannel, this feature is implied and curl then only uses the native CA
store.

Providing --ca-native multiple times has no extra effect.
Disable it again with \-\-no-ca-native.

Example:
.nf
 curl --ca-native https://example.com
.fi

See also \fI\-\-cacert\fP, \fI\-\-capath\fP and \fI-k, \-\-insecure\fP. Added in 8.2.0.
.IP "\-\-cacert <file>"
(TLS) Use the specified certificate file to verify the peer. The file may contain
multiple CA certificates. The certificate(s) must be in PEM format. Normally
curl is built to use a default file for this, so this option is typically used
to alter that default file.

curl recognizes the environment variable named \(aqCURL_CA_BUNDLE\(aq if it is set
and the TLS backend is not Schannel, and uses the given path as a path to a CA
cert bundle. This option overrides that variable.

The windows version of curl automatically looks for a CA certs file named
\(aqcurl\-ca\-bundle.crt\(aq, either in the same directory as curl.exe, or in the
Current Working Directory, or in any folder along your PATH.

(iOS and macOS only) If curl is built against Secure Transport, then this
option is supported for backward compatibility with other SSL engines, but it
should not be set. If the option is not set, then curl uses the certificates
in the system and user Keychain to verify the peer, which is the preferred
method of verifying the peer\(aqs certificate chain.

(Schannel only) This option is supported for Schannel in Windows 7 or later
(added in 7.60.0). This option is supported for backward compatibility with
other SSL engines; instead it is recommended to use Windows\(aq store of root
certificates (the default for Schannel).

If --cacert is provided several times, the last set value is used.

Example:
.nf
 curl --cacert CA-file.txt https://example.com
.fi

See also \fI\-\-capath\fP and \fI-k, \-\-insecure\fP.
.IP "\-\-capath <dir>"
(TLS) Use the specified certificate directory to verify the peer. Multiple paths can
be provided by separated with colon (":") (e.g. "path1:path2:path3"). The
certificates must be in PEM format, and if curl is built against OpenSSL, the
directory must have been processed using the c_rehash utility supplied with
OpenSSL. Using \fI\-\-capath\fP can allow OpenSSL\-powered curl to make SSL\-connections
much more efficiently than using \fI\-\-cacert\fP if the \fI\-\-cacert\fP file contains many
CA certificates.

If this option is set, the default capath value is ignored.

If --capath is provided several times, the last set value is used.

Example:
.nf
 curl --capath /local/directory https://example.com
.fi

See also \fI\-\-cacert\fP and \fI-k, \-\-insecure\fP.
.IP "\-E, \-\-cert <certificate[:password]>"
(TLS) Use the specified client certificate file when getting a file with HTTPS, FTPS
or another SSL\-based protocol. The certificate must be in PKCS#12 format if
using Secure Transport, or PEM format if using any other engine. If the
optional password is not specified, it is queried for on the terminal. Note
that this option assumes a certificate file that is the private key and the
client certificate concatenated. See \fI\-E, \-\-cert\fP and \fI\-\-key\fP to specify them
independently.

In the <certificate> portion of the argument, you must escape the character
\&":" as "\\:" so that it is not recognized as the password delimiter. Similarly,
you must escape the double quote character as \\" so that it is not recognized
as an escape character.

If curl is built against OpenSSL library, and the engine pkcs11 is available,
then a PKCS#11 URI (RFC 7512) can be used to specify a certificate located in
a PKCS#11 device. A string beginning with "pkcs11:" is interpreted as a
PKCS#11 URI. If a PKCS#11 URI is provided, then the \fI\-\-engine\fP option is set as
\&"pkcs11" if none was provided and the \fI\-\-cert\-type\fP option is set as "ENG" if
none was provided.

(iOS and macOS only) If curl is built against Secure Transport, then the
certificate string can either be the name of a certificate/private key in the
system or user keychain, or the path to a PKCS#12\-encoded certificate and
private key. If you want to use a file from the current directory, please
precede it with "./" prefix, in order to avoid confusion with a nickname.

(Schannel only) Client certificates must be specified by a path expression to
a certificate store. (Loading \fIPFX\fP is not supported; you can import it to a
store first). You can use "<store location>\\<store name>\\<thumbprint>"
to refer to a certificate in the system certificates store, for example,
\fI"CurrentUser\\MY\\934a7ac6f8a5d579285a74fa61e19f23ddfe8d7a"\fP. Thumbprint is
usually a SHA\-1 hex string which you can see in certificate details. Following
store locations are supported: \fICurrentUser\fP, \fILocalMachine\fP,
\fICurrentService\fP, \fIServices\fP, \fICurrentUserGroupPolicy\fP,
\fILocalMachineGroupPolicy\fP and \fILocalMachineEnterprise\fP.

If --cert is provided several times, the last set value is used.

Example:
.nf
 curl --cert certfile --key keyfile https://example.com
.fi

See also \fI\-\-cert\-type\fP, \fI\-\-key\fP and \fI\-\-key\-type\fP.
.IP "\-\-cert\-status"
(TLS) Verify the status of the server certificate by using the Certificate Status
Request (aka. OCSP stapling) TLS extension.

If this option is enabled and the server sends an invalid (e.g. expired)
response, if the response suggests that the server certificate has been
revoked, or no response at all is received, the verification fails.

This support is currently only implemented in the OpenSSL and GnuTLS backends.

Providing --cert-status multiple times has no extra effect.
Disable it again with \-\-no-cert-status.

Example:
.nf
 curl --cert-status https://example.com
.fi

See also \fI\-\-pinnedpubkey\fP.
.IP "\-\-cert\-type <type>"
(TLS) Set type of the provided client certificate. PEM, DER, ENG and P12 are
recognized types.

The default type depends on the TLS backend and is usually PEM, however for
Secure Transport and Schannel it is P12. If \fI\-E, \-\-cert\fP is a pkcs11: URI then ENG is
the default type.

If --cert-type is provided several times, the last set value is used.

Example:
.nf
 curl --cert-type PEM --cert file https://example.com
.fi

See also \fI-E, \-\-cert\fP, \fI\-\-key\fP and \fI\-\-key\-type\fP.
.IP "\-\-ciphers <list of ciphers>"
(TLS) Specifies which ciphers to use in the connection. The list of ciphers must
specify valid ciphers. Read up on SSL cipher list details on this URL:

https://curl.se/docs/ssl\-ciphers.html

If --ciphers is provided several times, the last set value is used.

Example:
.nf
 curl --ciphers ECDHE-ECDSA-AES256-CCM8 https://example.com
.fi

See also \fI\-\-tlsv1.3\fP, \fI\-\-tls13\-ciphers\fP and \fI\-\-proxy\-ciphers\fP.
.IP "\-\-compressed"
(HTTP) Request a compressed response using one of the algorithms curl supports, and
automatically decompress the content.

Response headers are not modified when saved, so if they are "interpreted"
separately again at a later point they might appear to be saying that the
content is (still) compressed; while in fact it has already been decompressed.

If this option is used and the server sends an unsupported encoding, curl
reports an error. This is a request, not an order; the server may or may not
deliver data compressed.

Providing --compressed multiple times has no extra effect.
Disable it again with \-\-no-compressed.

Example:
.nf
 curl --compressed https://example.com
.fi

See also \fI\-\-compressed\-ssh\fP.
.IP "\-\-compressed\-ssh"
(SCP SFTP) Enables built\-in SSH compression. This is a request, not an order; the server
may or may not do it.

Providing --compressed-ssh multiple times has no extra effect.
Disable it again with \-\-no-compressed-ssh.

Example:
.nf
 curl --compressed-ssh sftp://example.com/
.fi

See also \fI\-\-compressed\fP. Added in 7.56.0.
.IP "\-K, \-\-config <file>"
Specify a text file to read curl arguments from. The command line arguments
found in the text file are used as if they were provided on the command
line.

Options and their parameters must be specified on the same line in the file,
separated by whitespace, colon, or the equals sign. Long option names can
optionally be given in the config file without the initial double dashes and
if so, the colon or equals characters can be used as separators. If the option
is specified with one or two dashes, there can be no colon or equals character
between the option and its parameter.

If the parameter contains whitespace or starts with a colon (:) or equals sign
(=), it must be specified enclosed within double quotes ("like this"). Within
double quotes the following escape sequences are available: \\\\, \\", \\t, \\n, \\r
and \\v. A backslash preceding any other letter is ignored.

If the first non\-blank column of a config line is a \(aq#\(aq character, that line
is treated as a comment.

Only write one option per physical line in the config file. A single line is
required to be no more than 10 megabytes (since 8.2.0).

Specify the filename to \fI\-K, \-\-config\fP as minus "\-" to make curl read the file from
stdin.

Note that to be able to specify a URL in the config file, you need to specify
it using the \fI\-\-url\fP option, and not by simply writing the URL on its own
line. So, it could look similar to this:
.nf

url = "https://curl.se/docs/"

# \--\- Example file \--\-
# this is a comment
url = "example.com"
output = "curlhere.html"
user\-agent = "superagent/1.0"

# and fetch another URL too
url = "example.com/docs/manpage.html"
-O
referer = "http://nowhereatall.example.com/"
# \--\- End of example file \--\-
.fi

When curl is invoked, it (unless \fI\-q, \-\-disable\fP is used) checks for a default
config file and uses it if found, even when \fI\-K, \-\-config\fP is used. The default
config file is checked for in the following places in this order:

1) \fB"$CURL_HOME/.curlrc"\fP

2) \fB"$XDG_CONFIG_HOME/curlrc"\fP (Added in 7.73.0)

3) \fB"$HOME/.curlrc"\fP

4) Windows: \fB"%USERPROFILE%\\.curlrc"\fP

5) Windows: \fB"%APPDATA%\\.curlrc"\fP

6) Windows: \fB"%USERPROFILE%\\Application Data\\.curlrc"\fP

7) Non\-Windows: use getpwuid to find the home directory

8) On Windows, if it finds no \fI.curlrc\fP file in the sequence described above, it
checks for one in the same directory the curl executable is placed.

On Windows two filenames are checked per location: \fI.curlrc\fP and \fI_curlrc\fP,
preferring the former. Older versions on Windows checked for \fI_curlrc\fP only.

--config can be used several times in a command line

Example:
.nf
 curl --config file.txt https://example.com
.fi

See also \fI-q, \-\-disable\fP.
.IP "\-\-connect\-timeout <seconds>"
Maximum time in seconds that you allow curl\(aqs connection to take. This only
limits the connection phase, so if curl connects within the given period it
continues \- if not it exits.

This option accepts decimal values. The decimal value needs
to be provided using a dot (.) as decimal separator \- not the local version
even if it might be using another separator.

The connection phase is considered complete when the DNS lookup and requested
TCP, TLS or QUIC handshakes are done.

If --connect-timeout is provided several times, the last set value is used.

Examples:
.nf
 curl --connect-timeout 20 https://example.com
 curl --connect-timeout 3.14 https://example.com
.fi

See also \fI-m, \-\-max\-time\fP.
.IP "\-\-connect\-to <HOST1:PORT1:HOST2:PORT2>"
For a request intended for the "HOST1:PORT1" pair, connect to "HOST2:PORT2"
instead. This option is only used to establish the network connection. It does
NOT affect the hostname/port number that is used for TLS/SSL (e.g. SNI,
certificate verification) or for the application protocols.

\&"HOST1" and "PORT1" may be empty strings, meaning any host or any port number.
\&"HOST2" and "PORT2" may also be empty strings, meaning use the request\(aqs
original hostname and port number.

A hostname specified to this option is compared as a string, so it needs to
match the name used in request URL. It can be either numerical such as
\&"127.0.0.1" or the full host name such as "example.org".

--connect-to can be used several times in a command line

Example:
.nf
 curl --connect-to example.com:443:example.net:8443 https://example.com
.fi

See also \fI\-\-resolve\fP and \fI-H, \-\-header\fP.
.IP "\-C, \-\-continue\-at <offset>"
Resume a previous transfer from the given byte offset. The given offset is the
exact number of bytes that are skipped, counting from the beginning of the
source file before it is transferred to the destination. If used with uploads,
the FTP server command SIZE is not used by curl.

Use "\-C \-" to instruct curl to automatically find out where/how to resume the
transfer. It then uses the given output/input files to figure that out.

If --continue-at is provided several times, the last set value is used.

Examples:
.nf
 curl -C - https://example.com
 curl -C 400 https://example.com
.fi

See also \fI-r, \-\-range\fP.
.IP "\-b, \-\-cookie <data|filename>"
(HTTP) Pass the data to the HTTP server in the Cookie header. It is supposedly the
data previously received from the server in a "Set\-Cookie:" line. The data
should be in the format "NAME1=VALUE1; NAME2=VALUE2" or as a single filename.

When given a set of specific cookies and not a filename, it makes curl use the
cookie header with this content explicitly in all outgoing request(s). If
multiple requests are done due to authentication, followed redirects or
similar, they all get this cookie header passed on.

If no "=" symbol is used in the argument, it is instead treated as a filename
to read previously stored cookie from. This option also activates the cookie
engine which makes curl record incoming cookies, which may be handy if you are
using this in combination with the \fI\-L, \-\-location\fP option or do multiple URL
transfers on the same invoke.

If the filename is a single minus ("\-"), curl reads the contents from stdin.
If the filename is an empty string ("") and is the only cookie input, curl
activates the cookie engine without any cookies.

The file format of the file to read cookies from should be plain HTTP headers
(Set\-Cookie style) or the Netscape/Mozilla cookie file format.

The file specified with \fI\-b, \-\-cookie\fP is only used as input. No cookies are written
to that file. To store cookies, use the \fI\-c, \-\-cookie\-jar\fP option.

If you use the Set\-Cookie file format and do not specify a domain then the
cookie is not sent since the domain never matches. To address this, set a
domain in Set\-Cookie line (doing that includes subdomains) or preferably: use
the Netscape format.

Users often want to both read cookies from a file and write updated cookies
back to a file, so using both \fI\-b, \-\-cookie\fP and \fI\-c, \-\-cookie\-jar\fP in the same command
line is common.

If curl is built with PSL (\fBPublic Suffix List\fP) support, it detects and
discards cookies that are specified for such suffix domains that should not be
allowed to have cookies. If curl is \fInot\fP built with PSL support, it has no
ability to stop super cookies.

--cookie can be used several times in a command line

Examples:
.nf
 curl -b "" https://example.com
 curl -b cookiefile https://example.com
 curl -b cookiefile -c cookiefile https://example.com
 curl -b name=Jane https://example.com
.fi

See also \fI-c, \-\-cookie\-jar\fP and \fI-j, \-\-junk\-session\-cookies\fP.
.IP "\-c, \-\-cookie\-jar <filename>"
(HTTP) Specify to which file you want curl to write all cookies after a completed
operation. Curl writes all cookies from its in\-memory cookie storage to the
given file at the end of operations. Even if no cookies are known, a file is
created so that it removes any formerly existing cookies from the file. The
file uses the Netscape cookie file format. If you set the filename to a single
minus, "\-", the cookies are written to stdout.

The file specified with \fI\-c, \-\-cookie\-jar\fP is only used for output. No cookies are
read from the file. To read cookies, use the \fI\-b, \-\-cookie\fP option. Both options
can specify the same file.

This command line option activates the cookie engine that makes curl record
and use cookies. The \fI\-b, \-\-cookie\fP option also activates it.

If the cookie jar cannot be created or written to, the whole curl operation
does not fail or even report an error clearly. Using \fI\-v, \-\-verbose\fP gets a warning
displayed, but that is the only visible feedback you get about this possibly
lethal situation.

If --cookie-jar is provided several times, the last set value is used.

Examples:
.nf
 curl -c store-here.txt https://example.com
 curl -c store-here.txt -b read-these https://example.com
.fi

See also \fI-b, \-\-cookie\fP.
.IP "\-\-create\-dirs"
When used in conjunction with the \fI\-o, \-\-output\fP option, curl creates the necessary
local directory hierarchy as needed. This option creates the directories
mentioned with the \fI\-o, \-\-output\fP option combined with the path possibly set with
\fI\-\-output\-dir\fP. If the combined output filename uses no directory, or if the
directories it mentions already exist, no directories are created.

Created directories are made with mode 0750 on unix style file systems.

To create remote directories when using FTP or SFTP, try \fI\-\-ftp\-create\-dirs\fP.

Providing --create-dirs multiple times has no extra effect.
Disable it again with \-\-no-create-dirs.

Example:
.nf
 curl --create-dirs --output local/dir/file https://example.com
.fi


Libraries/curl/8.8.0/.brew/curl.rb
class Curl < Formula
  desc "Get a file from an HTTP, HTTPS or FTP server"
  homepage "https://curl.se"
  # Don't forget to update both instances of the version in the GitHub mirror URL.
  url "https://curl.se/download/curl-8.8.0.tar.bz2"
  mirror "https://github.com/curl/curl/releases/download/curl-8_8_0/curl-8.8.0.tar.bz2"
  mirror "http://fresh-center.net/linux/www/curl-8.8.0.tar.bz2"
  mirror "http://fresh-center.net/linux/www/legacy/curl-8.8.0.tar.bz2"
  sha256 "40d3792d38cfa244d8f692974a567e9a5f3387c547579f1124e95ea2a1020d0d"
  license "curl"

  livecheck do
    url "https://curl.se/download/"
    regex(/href=.*?curl[._-]v?(.*?)\.t/i)
  end

  head do
    url "https://github.com/curl/curl.git", branch: "master"

    depends_on "autoconf" => :build
    depends_on "automake" => :build
    depends_on "libtool" => :build
  end

  keg_only :provided_by_macos

  depends_on "pkg-config" => :build
  depends_on "brotli"
  depends_on "libidn2"
  depends_on "libnghttp2"
  depends_on "libssh2"
  depends_on "openldap"
  depends_on "openssl@3"
  depends_on "rtmpdump"
  depends_on "zstd"

  uses_from_macos "krb5"
  uses_from_macos "zlib"

  def install
    tag_name = "curl-#{version.to_s.tr(".", "_")}"
    if build.stable? && stable.mirrors.grep(/github\.com/).first.exclude?(tag_name)
      odie "Tag name #{tag_name} is not found in the GitHub mirror URL! " \
           "Please make sure the URL is correct."
    end

    system "./buildconf" if build.head?

    args = %W[
      --disable-debug
      --disable-dependency-tracking
      --disable-silent-rules
      --prefix=#{prefix}
      --with-ssl=#{Formula["openssl@3"].opt_prefix}
      --without-ca-bundle
      --without-ca-path
      --with-ca-fallback
      --with-secure-transport
      --with-default-ssl-backend=openssl
      --with-libidn2
      --with-librtmp
      --with-libssh2
      --without-libpsl
      --with-zsh-functions-dir=#{zsh_completion}
      --with-fish-functions-dir=#{fish_completion}
    ]

    args << if OS.mac?
      "--with-gssapi"
    else
      "--with-gssapi=#{Formula["krb5"].opt_prefix}"
    end

    system "./configure", *args
    system "make", "install"
    system "make", "install", "-C", "scripts"
    libexec.install "scripts/mk-ca-bundle.pl"
  end

  test do
    # Fetch the curl tarball and see that the checksum matches.
    # This requires a network connection, but so does Homebrew in general.
    filename = (testpath/"test.tar.gz")
    system "#{bin}/curl", "-L", stable.url, "-o", filename
    filename.verify_checksum stable.checksum

    system libexec/"mk-ca-bundle.pl", "test.pem"
    assert_predicate testpath/"test.pem", :exist?
    assert_predicate testpath/"certdata.txt", :exist?
  end
end
